src/main/java/com/aws/greengrass/clientdevices/auth/certificate/infra/BackgroundCertificateRefresh.java

```diff
@@ -190,77 +210,65 @@ private void refreshCertificateThingAttachment(Thing thing) {
-        Optional<String> certPem = pemStore.getPem(certificate.getCertificateId());
+    /**
+     * Closure around a Consumer that refreshes all the certificates attached to a thing. It records the certificates
+     * that have already been refreshed to avoid refreshing them again if they are attached to another thing and allows
+     * the enclosed value to be returned ba accessing the pair.getLeft()
+     */
+    private Pair<Set<String>, Consumer<Thing>> refreshCertificateValidity() {
+        Set<String> alreadyRefreshed = new HashSet<>();
+        VerifyIotCertificate verifyCertificate = useCases.get(VerifyIotCertificate.class);
 
-        if (!certPem.isPresent()) {
-            logger.atWarn().kv("certificateId", certificate.getCertificateId())
-                    .log("Tried to refresh certificate validity but its pem was not found");
-            return;
-        }
+        Consumer<Thing> callback = (Thing thing) -> {
+            Set<String> thingCertificateIds = thing.getAttachedCertificateIds().keySet();
 
-        VerifyIotCertificate useCase = useCases.get(VerifyIotCertificate.class);
-        useCase.apply(certPem.get());
-    }
+            thingCertificateIds.forEach(certificateId -> {
 
-    private boolean isNetworkDown() {
-        return networkState.getConnectionStateFromMqtt() == NetworkState.ConnectionState.NETWORK_DOWN;
-    }
+                if (alreadyRefreshed.contains(certificateId)) {
+                    return;
+                }
 
-    /**
-     * Represents which Things are still attached to the core device and which things are no longer attached.
-     */
-    private static class ThingAssociations {
-        private final Set<String> cloudThingNames;
-        private Set<String> attachedCertificateIds;
-        private List<Thing> attachedThings;
-        private List<Thing> detachedThings;
-
-        private ThingAssociations(Set<String> thingNamesAttachedToCore) {
-            this.cloudThingNames = thingNamesAttachedToCore;
-        }
+                Optional<String> certPem = pemStore.getPem(certificateId);
 
-        public static ThingAssociations create(Stream<Thing> thingsAttachedToCore) {
-            return new ThingAssociations(thingsAttachedToCore.map(Thing::getThingName).collect(Collectors.toSet()));
-        }
+                if (!certPem.isPresent()) {
+                    logger.atWarn()
+                            .kv("certificateId", certificateId)
+                            .kv("thing", thing.getThingName())
+                            .log("Tried to refresh certificate validity but its pem was not found");
+                    return;
+                }
 
-        /**
-         * Gets a stream of all the things that are still associated to the core.
-         */
-        public Stream<Thing> getThings() {
-            return attachedThings.stream();
-        }
+                alreadyRefreshed.add(certificateId);
```

Comment:
should we have this after L#215 `verifyCertificate.apply(certPem.get())`,  considering verifyCertificate usecase might fail.